Marlin/src/api/measurement/measureAtSwitch.cpp

```diff
@@ -24,29 +24,45 @@ int measureAtSwitch(const Endstop& endstop, float maxTravel, float& measurement)
       << endl;
     return -1;
   }
-  const float triggerPos = current_position[axis];
+  const float initialTriggerPos = current_position[axis];
+  float measurements[2] = {initialTriggerPos, initialTriggerPos};
 
-  // Retract slightly
-  if (retractFromSwitch(endstop)) {
-    return -1;
-  }
-  const float retractPos = current_position[axis];
-  const float retractDistance = abs(retractPos - triggerPos);
+  for (int i = 0; i < maxTouchCount; i++){
+    // Retract slightly
+    if (retractFromSwitch(endstop)) {
+      return -1;
+    }
+    const float retractPos = current_position[axis];
 
-  // Approach again, slowly
-  // NOTE: this gives us a more accurate reading
-  const auto slow = homing_feedrate[axis] / 6;
-  if (moveToEndstop(endstop, slow, 2 * retractDistance)) {
-    logError
-      << F("Unable to measure at ") << endstop.name
-      << F(" switch, switch did not trigger during second approach")
-      << endl;
-    return -1;
-  }
+    //limit retraction distance to the initial touch-retraction movement, don't let it continuously back off
+    const float retractDistance = abs(retractPos - measurements[i % 2]);
+
+    // Approach again, slowly
+    // NOTE: this gives us a more accurate reading
+    const auto slow = homing_feedrate[axis] / 6;
+    if (moveToEndstop(endstop, slow, 2 * retractDistance)) {
```

Comment:
changing the retractDistance above impacts max travel here.  
we go to initialTrigger position.  retract 1 mm  is retract position going to be 1mm from initialTrigger position then?

And we are iterating multiple times potentially if our measurements are not consistant for whatever reason does retractDistance monotonically decrease?  in which case the allowable movement 2*retractDistance also gets smaller.   Possibly risky.